File: examples/sample_pbasex.m
%% Load Integrals
gData = loadG('../gData/G_r512_k64_l4.h5',1);
%% Load Data
h_fid = fopen('high_counts.bin','r');
l_fid = fopen('low_counts.bin','r');
f_fid = fopen('fancy.bin','r');
data = fread(h_fid,[1023,1023],'double'); % load one image into a 2D array
data(:,:,2) = fread(l_fid,[1023,1023],'double'); % stack second image to make a 3D array
data(:,:,3) = fread(f_fid,[1023,1023],'double');
fclose(h_fid);
fclose(l_fid);
fclose(f_fid);
%% Do Inversion
x0 = 512;
y0 = 512;
fold = resizeFolded(foldQuadrant(data,x0,y0,[1,1,1,1]),512);
%%
out = pbasex(fold,gData,1);
%% Plot Inversion Data
figure;
subplot(2,1,1)
plot(out.E,out.IE(:,1:2)*diag(1./mean(out.IE(:,1:2)))) % normalize by image and plot
xlabel('Energy (eV)')
ylabel('Intensity (normalized)')
legend('High Counts','Low Counts')
title('Sample CPBASEX Inversion Data')
subplot(2,1,2)
plot(out.E,out.betas(:,:,1)) % plot betas versus energy for the first image
xlabel('Energy (eV)')
ylabel('Beta-value')
legend('beta-2','beta-4')
%%
figure;
subplot(1,3,1)
imagesc(data(:,:,3)); % image raw data
title('Raw Image')
subplot(1,3,2)
imagesc(out.recon(:,:,3)) % image fit data
title('Non-Inverted Fit')
subplot(1,3,3)
imagesc(out.inv(:,:,3)) % image inverted fit data
title('Inverted Fit')
caxis([0,1e2])

File: examples/sample_pbasex.py
import numpy as np
import matplotlib.pyplot as plt
from pbasex import pbasex, loadG
from quadrant import foldQuadrant, resizeFolded

# Load images
def load_image(path):
	return np.fromfile(path).reshape(1023,1023)
samples = ['low_counts', 'high_counts', 'fancy', 'fancier']
raw = np.dstack((load_image(sample+'.bin') for sample in samples))

# Fold images into quadrants
x0, y0 = 511, 511
quadrant_filter = [1,1,1,1]
folded = resizeFolded(foldQuadrant(raw, x0, y0, quadrant_filter), 512)

# Load inversion data
gData = loadG('G_r512_k512_l4.h5', 1)

# Apply the pBASEX algorithm
out = pbasex(folded, gData, make_images=True, alpha=4.1e-5)

# Plot some results
plt.figure(figsize=(12,9))
for i, sample in enumerate(samples):
	plt.subplot(4,4,4*i+1)
	plt.imshow(raw[:,:,i])
	plt.xticks([])
	plt.yticks([])
	plt.ylabel(sample)
	clim = plt.gci().get_clim()
	plt.clim(0,clim[1])
	if i==0:
		plt.title('Raw Image')
	plt.subplot(4,4,4*i+2)
	plt.plot(out['E'], out['IE'][:,i], 'k')
	plt.gca().ticklabel_format(axis='y', style='sci', scilimits=(-2,2))
	if i==3:
		plt.xlabel('Energy (eV)')
	plt.gca().twinx()
	plt.plot(out['E'], out['betas'][:,:,i], '.', markersize=5)
	if i==0:
		plt.text(-3, 3.5, 'counts per eV', size='small')
		plt.text(12, 3.5, 'beta', size='small')
		plt.text(3.5, 3.25, 'I(E), ', color='black', size='large')
		plt.text(6, 3.25, 'B2', color='blue', size='large')
		plt.text(7.5, 3.25, ', ', color='black', size='large')
		plt.text(8, 3.25, 'B4', color='green', size='large')
	plt.ylim(-1,3)
	plt.subplot(4,4,4*i+3)
	plt.imshow(out['fit'][:,:,i]/4)
	plt.xticks([])
	plt.yticks([])
	plt.clim(0,clim[1])
	if i==0:
		plt.title('Fitted Image')
	plt.subplot(4,4,4*i+4)
	plt.imshow(out['inv'][:,:,i]/4)
	plt.xticks([])
	plt.yticks([])
	plt.clim(0,clim[1]/10)
	if i==0:
		plt.title('Inverted Image')
plt.show()